src/processor/tree/mod.rs

```diff
@@ -26,30 +26,31 @@ pub struct TreeProcessor {
     /// The internal merkle tree.
     tree: TreeWrapper,
     /// The stored state snapshot.
-    snapshot: Arc<Mutex<StateSnapshot>>,
+    snapshot: Arc<Mutex<InnerDB>>,
 }
 
 impl TreeProcessor {
-    pub async fn new(db_path: PathBuf, snapshot: Arc<Mutex<StateSnapshot>>) -> Result<Self> {
-        // If database directory already exists, we try to restore the latest state.
-        // The state contains the last processed block and a mapping of index to key
-        // values, if a state file does not exist, we simply use the defaults instead.
-        let should_restore_state = db_path.exists();
-        if should_restore_state {
-            tracing::info!("Loading previous state file...");
-            let new_state = StateSnapshot::read(&db_path.join(STATE_FILE_NAME))
-                .expect("state file is malformed");
-            *snapshot.lock().await = new_state;
-        } else {
-            tracing::info!("No existing database found, starting from genesis...");
-        };
+    pub async fn new(db_path: PathBuf) -> Result<Self> {
+        // If database directory already exists, we try to restore the
+        // latest state from a database inside of it.  The state
+        // contains the last processed block and a mapping of index to
+        // key values.
+        let inner_db_path = db_path.join(INNER_DB_NAME);
+        let init = !db_path.exists() || !inner_db_path.exists();
+        if init {
+            tracing::info!("No existing snapshot found, starting from genesis...");
+        }
 
-        // Extract `index_to_key_map` from state snapshot.
-        let index_to_key_map = snapshot.lock().await.index_to_key_map.clone();
-        let tree = TreeWrapper::new(&db_path, index_to_key_map)?;
+        let new_state = InnerDB::new(inner_db_path.clone())?;
+        let snapshot = Arc::new(Mutex::new(new_state));
+        let tree = TreeWrapper::new(&db_path, snapshot.clone(), init).await?;
 
         Ok(Self { tree, snapshot })
     }
+
+    pub fn get_snapshot(&self) -> Arc<Mutex<InnerDB>> {
+        self.snapshot.clone()
+    }
```

Comment:
Well, `TreeProcessor` is instantiated in 2 places - if it needs a database, why shouldn't it create/open it? And if it has a database, why shouln't it provide it? OK, it shouldn't provide it because it shouldn't share it - here we go again... :-)

Reply:
Ah wait that's right, and we also shouldn't open the inner database before we have had the chance to open the outer one. We can probably keep this solution for now then. :see_no_evil: